XMonad/Layout/PerScreen.hs

```diff
@@ -1,75 +1,119 @@
--- Copyright   :  (c) Edward Z. Yang
+-- Description :  Configure layouts based on the screen rectangle.
+-- Copyright   :  (c) Brandon S. Allbery KF8NH
 -- License     :  BSD-style (see LICENSE)
 --
--- Maintainer  :  <ezyang@cs.stanford.edu>
+-- Maintainer  :  <allbery.b@gmail.com>
 -- Stability   :  unstable
 -- Portability :  unportable
 --
--- Configure layouts based on the width of your screen; use your
--- favorite multi-column layout for wide screens and a full-screen
--- layout for small ones.
+-- Configure layouts based on the screen rectangle passed to the layout.
+-- This gives you true per-screen functionality.
+--
+-- The old PerScreen is now X.L.ByWidth. We re-export it deprecated for
+-- backward compatibility.
 -----------------------------------------------------------------------------
 
 module XMonad.Layout.PerScreen
     ( -- * Usage
       -- $usage
-      PerScreen,
+      OnScreen,
+      onScreen,
+      onScreens,
+      -- * Deprecated
+      -- $deprecated
+      BW.PerScreen,
       ifWider
     ) where
 
 import XMonad
 import qualified XMonad.StackSet as W
 
-import XMonad.Prelude (fromMaybe)
+import XMonad.Prelude (fromMaybe, fi)
+
+import qualified XMonad.Layout.ByWidth as BW
+
+import Data.List (find)
 
 -- $usage
 -- You can use this module by importing it into your ~\/.xmonad\/xmonad.hs file:
 --
--- > import XMonad.Layout.PerScreen
+-- > import XMonad.Layout.OnScreen
 --
 -- and modifying your layoutHook as follows (for example):
 --
--- > layoutHook = ifWider 1280 (Tall 1 (3/100) (1/2) ||| Full) Full
+-- > layoutHook = onScreen 1 (Tall 1 (3/100) (1/2) ||| Full) Full
 --
 -- Replace any of the layouts with any arbitrarily complicated layout.
--- ifWider can also be used inside other layout combinators.
+-- 'onScreen' can also be used inside other layout combinators, although the
+-- result may be confusing.
+
+-- | Specify a layout to run on a given screen.
+onScreen :: (LayoutClass l1 a, LayoutClass l2 a)
+         => ScreenId -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreen s = onScreens [s]
+
+-- | Specify a layout to run on a list of screens.
+--   Note that this works by 'ScreenId'. It has a 'Num' instance, so literal
+--   screen numbers will work as expected, but if you use a binding you need
+--   to use the 'S' constructor.
+onScreens :: (LayoutClass l1 a, LayoutClass l2 a)
+          => [ScreenId] -> l1 a -> l2 a -> OnScreen l1 l2 a
+onScreens ss l1 l2 = OnScreen ss l1 l2 False -- @@@ is this right?
+
+data OnScreen l1 l2 a = OnScreen [ScreenId] (l1 a) (l2 a) Bool
+    deriving (Read, Show)
+
+instance (LayoutClass l1 a, LayoutClass l2 a, Show a) => LayoutClass (OnScreen l1 l2) a where
+    runLayout (W.Workspace i p@(OnScreen ss l1 l2 _) ms) r = do
+        -- this sucks.
+        -- You might think we could use the existing screen detail, but we may be
+        -- invoked from 'rescreen' in which case only 'rescreen' and 'windows'
```

Comment:
But `windows` does `modify (\s -> s { windowset = ws })` pretty early (definitely before running any layouts) so why can't you just `withWindowSet` like X.L.NoBorders does (it also has logic to detect which screen is which based on the rect passed)?

Reply:
By not using the windowset this won't work with XMonad.Layout.LayoutScreens.

Reply:
I think `LayoutScreens` should be deprecated in favor of using `xrandr`: applications can't ever work with `LayoutScreens`.